File: tenk-backend/src/main/java/com/hjson/tenk/domain/amount/SpendCategory.java
package com.hjson.tenk.domain.amount;

import com.hjson.tenk.common.exception.BusinessException;
import com.hjson.tenk.common.exception.ErrorCode;

/**
 * 지출 카테고리 (고정 9종). {@code amount.category} 컬럼에는 이 enum 의 {@code name()}(코드)이 저장된다.
 *
 * <p>저장·전송은 안정적인 코드({@code FOOD}), 표시는 한글 {@code label}(식비) — 라벨을 바꿔도 DB
 * 마이그레이션이 필요 없다. 클라이언트는 코드를 받아 라벨·아이콘으로 매핑한다
 * ({@code lib/presentation/amount/spend_category.dart}).
 *
 * <p>엔티티 컬럼은 {@code VARCHAR + @Enumerated(EnumType.STRING)} 이다 (2026-07-30 전환).
 * 예전엔 "쓰기는 엄격, 읽기는 관대" 로 String 필드를 뒀는데, 이는 카테고리 검증 도입(2026-07-11)
 * <b>이전에</b> 저장된 자유 텍스트 row 를 읽을 때 enum 매핑 크래시를 피하려던 것이었다. 그 레거시 row 를
 * {@code ETC} 로 접는 마이그레이션을 끝냈으므로 관대할 이유가 사라졌다.
 *
 * <p>외부 입력(JSON)을 enum 으로 바꾸는 지점은 {@link #from(String)} 이고, 호출은 엔티티의 정적 팩토리
 * 안에서만 한다 — 요청 DTO 필드를 이 타입으로 올리지 말 것. Jackson 이 먼저 파싱에 실패하면
 * {@link ErrorCode#AMOUNT_CATEGORY_INVALID}(한국어 메시지) 대신 범용 400 이 나간다.
 */
public enum SpendCategory {
    FOOD("식비"),
    TRANSPORT("교통비"),
    SHOPPING("쇼핑"),
    LEISURE("여가"),
    HEALTH("건강"),
    EDUCATION("교육"),
    EVENT("경조사"),
    LIVING("생활비"),
    ETC("기타");

    private final String label;

    SpendCategory(String label) {
        this.label = label;
    }

    public String getLabel() {
        return label;
    }

    public static boolean isValidCode(String code) {
        if (code == null) {
            return false;
        }
        for (SpendCategory category : values()) {
            if (category.name().equals(code)) {
                return true;
            }
        }
        return false;
    }

    /// 코드 문자열 → enum. 9종 밖이면 {@link ErrorCode#AMOUNT_CATEGORY_INVALID}.
    ///
    /// null·공백은 예외가 아니라 **null 을 반환**한다 — "카테고리를 안 보냄"과 "이상한 값을 보냄"은
    /// 다른 에러(전자는 `AMOUNT_CATEGORY_CONTENT_REQUIRED`)라서, 미입력 판정은 호출자에게 남긴다.
    public static SpendCategory from(String code) {
        if (code == null || code.isBlank()) {
            return null;
        }
        for (SpendCategory category : values()) {
            if (category.name().equals(code)) {
                return category;
            }
        }
        throw new BusinessException(ErrorCode.AMOUNT_CATEGORY_INVALID);
    }
}

File: tenk_app/lib/presentation/amount/spend_category.dart
import 'package:flutter/material.dart';

/// 지출 카테고리 (고정 9종). 백엔드 `SpendCategory` enum 과 코드가 1:1 매칭된다.
///
/// 전송·저장은 안정적인 [code](예: `FOOD`), 표시는 [label](식비)·[icon]. 아이콘은 Material 벡터
/// (`IconData`) 라 색이 박혀있지 않고 렌더 시점에 테마 색으로 칠해진다 — 나중에 챌린지별 색을
/// 부여해도 자유롭게 대응된다. 여기가 code→label/icon 매핑의 진실의 원천.
@immutable
class SpendCategory {
  const SpendCategory({
    required this.code,
    required this.label,
    required this.icon,
  });

  final String code;
  final String label;
  final IconData icon;
}

/// 표시 순서 = 목록 순서. 백엔드 enum 선언 순서와 맞춰둔다.
const List<SpendCategory> kSpendCategories = [
  SpendCategory(code: 'FOOD', label: '식비', icon: Icons.restaurant),
  SpendCategory(code: 'TRANSPORT', label: '교통비', icon: Icons.directions_bus),
  SpendCategory(code: 'SHOPPING', label: '쇼핑', icon: Icons.shopping_bag),
  SpendCategory(code: 'LEISURE', label: '여가', icon: Icons.local_activity),
  SpendCategory(code: 'HEALTH', label: '건강', icon: Icons.health_and_safety),
  SpendCategory(code: 'EDUCATION', label: '교육', icon: Icons.school),
  SpendCategory(code: 'EVENT', label: '경조사', icon: Icons.card_giftcard),
  SpendCategory(code: 'LIVING', label: '생활비', icon: Icons.home),
  SpendCategory(code: 'ETC', label: '기타', icon: Icons.more_horiz),
];

/// 미매칭(옛 자유 텍스트·null)일 때의 폴백 — 기타.
const SpendCategory _fallbackCategory = SpendCategory(
  code: 'ETC',
  label: '기타',
  icon: Icons.more_horiz,
);

/// 코드로 카테고리를 찾는다. 9종에 없으면 [_fallbackCategory](기타) 로 폴백해
/// 화면이 깨지지 않게 한다 (검증 도입 이전에 저장된 자유 텍스트 대응).
SpendCategory spendCategoryForCode(String? code) {
  for (final category in kSpendCategories) {
    if (category.code == code) return category;
  }
  return _fallbackCategory;
}
